File: Swift/프로그래머스/2/12978. 배달/배달.swift
import Foundation

func solution(_ N:Int, _ road:[[Int]], _ k:Int) -> Int {
    var adjList: [[Node]] = .init(repeating: [], count: N+1)
    
    for edge in road {
        adjList[edge[0]].append(Node(edge[1], edge[2]))
        adjList[edge[1]].append(Node(edge[0], edge[2]))
    }
    
    var dist: [Int] = .init(repeating: .max, count: N+1)
    
    var pq: [Node] = [Node(1, 0)]
    dist[1] = 0
    
    while (!pq.isEmpty) {
        var now = pq.removeFirst()
        
        guard dist[now.dest] >= now.cost else { continue }
        
        for next in adjList[now.dest] {
            if dist[next.dest] > (now.cost + next.cost) {
                dist[next.dest] = now.cost + next.cost
                pq.append(Node(next.dest, dist[next.dest]))
            }
        }
    }
    
    var answer = 0
    
    for i in 1...N {
        if dist[i] <= k {
            answer += 1
        }
    }

    return answer
}

struct Node {
    let dest: Int
    let cost: Int
    
    init(_ dest: Int, _ cost: Int) {
        self.dest = dest
        self.cost = cost
    }
}

File: Java/프로그래머스/2/12978. 배달/배달.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.ArrayDeque;

class Solution {
    static class Node {
        int dest, cost;
        
        Node(int dest, int cost) {
            this.dest = dest;
            this.cost = cost;
        }
    }
    
    
    public int solution(int N, int[][] road, int K) {
        ArrayList<Node>[] map = new ArrayList[N+1];
        for (int i = 1; i<= N; i++) {
            map[i] = new ArrayList<>();
        }
        
        for (int[] item: road) {
            map[item[0]].add(new Node(item[1], item[2]));
            map[item[1]].add(new Node(item[0], item[2]));
        }
        
        int[] costList = new int[N+1];
        Arrays.fill(costList, Integer.MAX_VALUE);
        ArrayDeque<Node> queue = new ArrayDeque<>();
        queue.add(new Node(1, 0));
        costList[1] = 0;
        
        while(!queue.isEmpty()) {
            Node now = queue.poll();
            
            if (costList[now.dest] < now.cost)
                continue;
            
            for (Node connected : map[now.dest]) {
                if (costList[connected.dest] > now.cost + connected.cost) {
                    costList[connected.dest] = now.cost + connected.cost;
                    queue.add(new Node(connected.dest, costList[connected.dest]));
                }
            }
        }
        
        int answer = 0;
        for (int i = 1; i < costList.length; i++) {
            if (costList[i] <= K)
                answer++;
        }

        return answer;
    }
}